hydromt/models/model_mesh.py

```diff
@@ -253,33 +293,163 @@ def set_mesh(
+                        # Re-define _mesh
+                        grids = xr.merge(grids)
+                        self._mesh = xu.UgridDataset(grids)
+            # Check again mesh_names, could have changed if overwrite_grid=True
+            if grid_name in self.mesh_names:
+                for dvar in data.data_vars:
+                    if dvar in self._mesh:
+                        self.logger.warning(f"Replacing mesh parameter: {dvar}")
+                    self._mesh[dvar] = data[dvar]
+            else:
+                # We are potentially adding a new grid without any data variables
+                self._mesh = xu.UgridDataset(
+                    xr.merge([self.mesh.ugrid.to_dataset(), data.ugrid.to_dataset()])
+                )
+            # Restore crs
+            for grid in self._mesh.ugrid.grids:
+                grid.set_crs(crs)
+
+        # update related geoms if necessary: region
+        if overwrite_grid or new_grid:
+            # add / updates region
+            self._geoms.pop("region", None)
+            self.region
+
+    def get_mesh(
+        self, grid_name: str, include_data: bool = False
+    ) -> Union[xu.UgridDataArray, xu.UgridDataset]:
+        """
+        Return a specific grid topology from mesh based on grid_name.
+
+        If include_data is True, the data variables for that specific
+        grid are also included.
+
+        Parameters
+        ----------
+        grid_name : str
+            Name of the grid to return.
+        include_data : bool, optional
+            If True, also include data variables, by default False.
 
-    def read_mesh(self, fn: str = "mesh/mesh.nc", **kwargs) -> None:
+        Returns
+        -------
+        uds: Union[xu.UgridDataArray, xu.UgridDataset]
+            Grid topology with or without data variables.
+        """
+        if self.mesh is None:
+            raise ValueError("Mesh is not set, please use set_mesh first.")
+        if grid_name not in self.mesh_names:
+            raise ValueError(f"Grid {grid_name} not found in mesh.")
+        if include_data:
+            grid = self.mesh_grids[grid_name]
+            uds = xu.UgridDataset(
+                grid.to_dataset(optional_attributes=True)
+            )  # FIXME: would be nice if always write all attributes https://github.com/Deltares/xugrid/issues/140
+            uds.ugrid.grid.set_crs(grid.crs)
+            # Look for data_vars that are defined on grid_name
+            for var in self.mesh.data_vars:
+                if hasattr(self.mesh[var], "ugrid"):
+                    if self.mesh[var].ugrid.grid.name == grid_name:
+                        uds[var] = self.mesh[var]
+                # additionnal topology properties
+                elif var.startswith(grid_name):
+                    uds[var] = self.mesh[var]
+                # else is global property (not grid specific)
+
+            return uds
+
+        else:
+            return self.mesh_grids[grid_name]
```

Comment:
Hi @hboisgon, I slitely modified this function to return only xu.UgridDataset. Before it could return `Ugrid1d`/`Ugrid2d` (if `include_data` is True, or `UgridDataset` if `include_data` is False). I find that a little inconsistant. Considering when we call self.mesh it returns a `UgridDataset`, I think `get_mesh` returning `UgridDataset` make sense. Then the downstream actions for getting the grid `UgridDataset.grid` stays the same with or without data.  Could you review this and check if this does what you would like it to? 

```suggestion

    def get_mesh(self, grid_name: str, include_data: bool = False) -> xu.UgridDataset:
        """
        Return a specific ugrid dataset from mesh based on grid_name.

        If include_data is True, the data variables (including optional attribtues)
        for that specific grid are also included.

        Parameters
        ----------
        grid_name : str
            Name of the grid to return.
        include_data : bool, optional
            If True, also include data variables, by default False.

        Returns
        -------
        uds: xu.UgridDataset
            UGrid dataset with or without data variables.
        """
        if self.mesh is None:
            raise ValueError("Mesh is not set, please use set_mesh first.")
        if grid_name not in self.mesh_names:
            raise ValueError(f"Grid {grid_name} not found in mesh.")
        if include_data:
            grid = self.mesh_grids[grid_name]
            # FIXME xugrid #138 #140
            uds = xu.UgridDataset(grid.to_dataset(optional_attributes=True))
            uds.ugrid.grid.set_crs(grid.crs)
            # Look for data_vars that are defined on grid_name
            for var in self.mesh.data_vars:
                if hasattr(self.mesh[var], "ugrid"):
                    if self.mesh[var].ugrid.grid.name == grid_name:
                        uds[var] = self.mesh[var]
                # additionnal topology properties
                elif var.startswith(grid_name):
                    uds[var] = self.mesh[var]
                # else is global property (not grid specific)
        else:
            uds = xu.UgridDataset(self.mesh_grids[grid_name].to_dataset(optional_attributes = True))

        return uds
```